### ISSUE TITLE
I18n Cldr DatesReader Cache variable parsedFormats is sometimes empty

### ISSUE THREAD

**Issue by neos-bot** [neos/flow-development-collection#564]
Jira issue originally created by user timo.kiefer:

We have that problem about every 6 month, that the parsedFormats variable in DatesReader Cache is empty. We don't how this error occured. Our work around is to remove the whole caches of DatesReader.

``` java
Catchable Fatal Error: Argument 2 passed to TYPO3\Flow\I18n\Formatter\DatetimeFormatter*Original::doFormattingWithParsedFormat() must be of the type array, null given, called in /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow_Object_Classes/TYPO3_Flow_I18n_Formatter_DatetimeFormatter.php on line 108 and defined in /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow_Object_Classes/TYPO3_Flow_I18n_Formatter*DatetimeFormatter.php line 155

#0 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/TYPO3_Flow_I18n_Formatter*DatetimeFormatter.php(155): TYPO3\Flow\Error\ErrorHandler->handleError(4096, 'Argument 2 pass...', '/var/www/SETUPI...', 155, Array) 
#1 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/TYPO3_Flow_I18n_Formatter_DatetimeFormatter.php(108): TYPO3\Flow\I18n\Formatter\DatetimeFormatter_Original->doFormattingWithParsedFormat(Object(DateTime), NULL, Array) #2 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow_Object_Classes/TYPO3_Flow_I18n_Formatter_DatetimeFormatter.php(68): TYPO3\Flow\I18n\Formatter\DatetimeFormatter*Original->formatDate(Object(DateTime), Object(TYPO3\Flow\I18n\Locale), 'default') 
#3 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_ProductController.php(793): TYPO3\Flow\I18n\Formatter\DatetimeFormatter*Original->format(Object(DateTime), Object(TYPO3\Flow\I18n\Locale), Array) 
#4 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_ProductController.php(281): DGP\PIG\ViewHelpers\Widget\Controller\ProductController_Original->getTooltip() #5 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow_Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_ProductController.php(431): DGP\PIG\ViewHelpers\Widget\Controller\ProductController*Original->transformObject() 
#6 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_ProductController.php(408): DGP\PIG\ViewHelpers\Widget\Controller\ProductController*Original->transformChildren(Array) 
#7 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_AbstractWidgetController.php(186): DGP\PIG\ViewHelpers\Widget\Controller\ProductController*Original->transformObject() 
#8 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_AbstractWidgetController.php(171): DGP\PIG\ViewHelpers\Widget\Controller\AbstractWidgetController*Original->transformValue() 
#9 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_AbstractWidgetController.php(124): DGP\PIG\ViewHelpers\Widget\Controller\AbstractWidgetController*Original->transformValue() 
#10 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget_Controller_ProductController.php(164): DGP\PIG\ViewHelpers\Widget\Controller\AbstractWidgetController*Original->indexAction() 
#11 [internal function]: DGP\PIG\ViewHelpers\Widget\Controller\ProductController_Original->indexAction() 
#12 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/TYPO3_Flow_Mvc_Controller_ActionController.php(373): call_user_func*array(Array, Array) 
#13 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/TYPO3_Flow_Mvc_Controller_ActionController.php(169): TYPO3\Flow\Mvc\Controller\ActionController*Original->callActionMethod() 
#14 /var/www/SETUPIM/Packages/Framework/TYPO3.Fluid/Classes/TYPO3/Fluid/Core/Widget/AbstractWidgetController.php(53): TYPO3\Flow\Mvc\Controller\ActionController_Original->processRequest(Object(TYPO3\Flow\Mvc\ActionRequest), Object(TYPO3\Flow\Http\Response)) 
#15 /var/www/SETUPIM/Packages/Framework/TYPO3.Fluid/Classes/TYPO3/Fluid/Core/Widget/AbstractWidgetViewHelper.php(198): TYPO3\Fluid\Core\Widget\AbstractWidgetController->processRequest(Object(TYPO3\Flow\Mvc\ActionRequest), Object(TYPO3\Flow\Http\Response)) 
#16 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/DGP_PIG_ViewHelpers_Widget*ProductViewHelper.php(42): TYPO3\Fluid\Core\Widget\AbstractWidgetViewHelper->initiateSubRequest() 
#17 [internal function]: DGP\PIG\ViewHelpers\Widget\ProductViewHelper_Original->render(Object(TYPO3\Flow\Persistence\Doctrine\QueryResult), Object(DGP\PIG\Domain\Model\Node), Object(DateTime), Object(DGP\PIG\Domain\Model\Type), NULL, NULL, NULL, 'editing') 
#18 /var/www/SETUPIM/Packages/Framework/TYPO3.Fluid/Classes/TYPO3/Fluid/Core/ViewHelper/AbstractViewHelper.php(270): call*user_func*array(Array, Array) #19 /var/www/SETUPIM/Packages/Framework/TYPO3.Fluid/Classes/TYPO3/Fluid/Core/Widget/AbstractWidgetViewHelper.php(102): TYPO3\Fluid\Core\ViewHelper\AbstractViewHelper->callRenderMethod() 
#20 /var/www/SETUPIM/Data/Temporary/Production/Cache/Code/Flow*Object_Classes/TYPO3_Fluid_Core_Parser_SyntaxTree*ViewHelperNode.php(141): TYPO3\Fluid\Core\Widget\AbstractWidgetViewHelper->initializeArgumentsAndRender() 
#21 /var/www/SETUPIM/Packages/Framework/TYPO3.Fluid/Classes/TYPO3/Fluid/Core/Parser/SyntaxTree/AbstractNode.php(51): TYPO3\Fluid\Core\Parser\SyntaxTree\ViewHelperNode_Original->evaluate(Object(TYPO3\Fluid\Core\Rendering\RenderingContext)) 
#22 ...
```

 Jira-URL: https://jira.neos.io/browse/FLOW-277

**Comment by baschny** [comment]
Funny enough, we still struggle with this exact same bug in 2020, happens every couple of weeks in one of our production sites. Now we figured it out, but we need to wait until it happens again to be able to debug something.

**Comment by albe** [comment]
Oh my, that would be awesome to get some more info on this moon-phase bug. The thing that particularly strikes me as odd, is that the method that fills in the above mentioned "argument 2" is `parseFormatFromCldr` and (as of 6.0) has a strict `: array` return typehint. Also, there is no obvious execution path that could lead to the method returning `null`...

The only reasonable thing that could lead to this error above is, iff the `parsedFormatsIndices` cache diverged from the `parsedFormats` cache such that the latter would be empty, while the former still contained values. So basically, this comes down to a race condition, where flushing caches is being executed and `parsedFormats` is already cleaned but not `parsedFormatsIndices` and a request comes in. As both share the same TTL this can happen every time the caches time out, which would explain why the bug occurs every couple months/weeks. Did you configure those caches TTL by chance? Or do you have deployments every couple weeks (which trigger a cache flush)?

A workaround could be to make the check at
https://github.com/neos/flow-development-collection/blob/5.3/Neos.Flow/Classes/I18n/Cldr/Reader/DatesReader.php#L264
also check for existence of `parsedFormats`.

Something along the lines of:
```php
        $parsedFormatIndex = $this->parsedFormatsIndices[(string)$locale][$formatType][$formatLength] ?? '';
        if (isset($this->parsedFormats[$parsedFormatIndex])) {
            return $this->parsedFormats[$parsedFormatIndex];
        }
```

**Comment by baschny** [comment]
I have not configured that caches, so they use the defaults (no TTL, file system backend etc). A deployment flushes the cache. The problem only ssems to happen when there is a longer phase without manually cache flushing.

The proposed check would indeed make it safer.

The exact flow to the race condition is unclear. But having two separate cache entries which depend on each other implicitly is always a "bad idea". Maybe this code needs to be refactored to avoid this?

**Comment by albe** [comment]
> when there is a longer phase without manually cache flushing.

That's odd then, as that doesn't fit to the reasoning above, which would mean there's another possible cause 🤔

> But having two separate cache entries which depend on each other implicitly is always a "bad idea". Maybe this code needs to be refactored to avoid this?

Indeed. And that was also my first intent, but there is a reason for this split cache: it's possible to parse custom formats and those should also be cached, but are not indexed by locale, format type and length. So either the caches are split (while containing the same information) and the index cache removed or things are left as is and the race condition avoided otherwise.

Would be cool if you could check out the above code change once the bug reappears for you. I'll provide a PR already, since this check is easy and makes sense in any circumstance.

**Comment by baschny** [comment]
The bug appeared live again today, and the above fix you provided fixes the issue (i.e. no more Exception shown). So my suggestion would be to at least add this fix in current supported versions (we would love to see it in Flow 4.3 as this is the current project we are using this), and think about a better solution in general for the future separately.

**Comment by baschny** [comment]
I just created a MR with the fix included, maybe someone can take a look at it: 
https://github.com/neos/flow-development-collection/pull/1899

**Comment by baschny** [comment]
This fix for the `DatesReader` is already merged and released with Flow 5.3.12, 6.0.10 and 6.1.3 and later. It is already merged in branch 4.3, but there was no release yet since 4.3.23 unfortunately. Probably never will? @kitsunet 

Other than that, I just discovered on my site that the problem also hits the `NumbersReader` class, as it has a similar (exactly the same) code and the fix wasn't applied to that. I will open a new MR for that, hopefull to get it merged also in 4.3 (and maybe do a release 4.3.24, since there is only this and some other tiny bugfix included).

**Comment by albe** [comment]
closing as of the two PRs and the bugfix release done back then :)
